lib/runner.js

```diff
@@ -377,6 +378,15 @@ Runner.prototype.hook = function(name, fn) {
       }
       if (err) {
         if (err instanceof Pending) {
+          if (name === HOOK_TYPE_AFTER_EACH || name === HOOK_TYPE_AFTER_ALL) {
+            utils.deprecate(
+              format(
+                "skipping a test within an '%s' hook is DEPRECATED and will be removed from a future version of Mocha. " +
+                  "Use 'return' instead to exit hook abortively.",
+                name
+              )
+            );
+          }
           if (name === HOOK_TYPE_BEFORE_EACH || name === HOOK_TYPE_AFTER_EACH) {
             self.test.pending = true;
```

Comment:
OK, so given the discussion here, I don't think this should be happening for "after each".  do you want to change this, and add a test?

Reply:
...or just leave the deprecation message and change it later?  it's not clear to me what this would break.

Reply:
@boneskull yes, I will change this. To be consistent ' after` should be changed also.

Reply:
ok, just deprecation message, done.